sycl/doc/extensions/proposed/sycl_ext_oneapi_raw_kernel.asciidoc

```diff
@@ -0,0 +1,563 @@
+= sycl_ext_oneapi_raw_kernel
+
+:source-highlighter: coderay
+:coderay-linenums-mode: table
+
+// This section needs to be after the document title.
+:doctype: book
+:toc2:
+:toc: left
+:encoding: utf-8
+:lang: en
+:dpcpp: pass:[DPC++]
+
+// Set the default source code type in this document to C++,
+// for syntax highlighting purposes.  This is needed because
+// docbook uses c++ and html5 uses cpp.
+:language: {basebackend@docbook:c++:cpp}
+
+
+== Notice
+
+[%hardbreaks]
+Copyright (C) 2023-2023 Intel Corporation.  All rights reserved.
+
+Khronos(R) is a registered trademark and SYCL(TM) and SPIR(TM) are trademarks
+of The Khronos Group Inc.
+OpenCL(TM) is a trademark of Apple Inc. used by permission by Khronos.
+
+
+== Contact
+
+To report problems with this extension, please open a new issue at:
+
+https://github.com/intel/llvm/issues
+
+
+== Dependencies
+
+This extension is written against the SYCL 2020 revision 8 specification.
+All references below to the "core SYCL specification" or to section numbers in
+the SYCL specification refer to that revision.
+
+This extension also depends on the following other SYCL extensions:
+
+* link:../experimental/sycl_ext_oneapi_properties.asciidoc[
+  sycl_ext_oneapi_properties]
+* link:../proposed/sycl_ext_oneapi_free_function_queries.asciidoc[
+  sycl_ext_oneapi_free_function_queries]
+
+
+== Status
+
+This is a proposed extension specification, intended to gather community
+feedback.
+Interfaces defined in this specification may not be implemented yet or may be
+in a preliminary state.
+The specification itself may also change in incompatible ways before it is
+finalized.
+*Shipping software products should not rely on APIs defined in this
+specification.*
+
+
+== Overview
+
+This extension introduces a new way to define a kernel as a simple C++
+function, where the kernel arguments are parameters to the function.
+This is different from standard SYCL kernels, where the kernel arguments are
+either captures of a lambda expression or member variables of a callable
+object.
+
+The primary motivation for this extension is the
+link:../proposed/sycl_ext_oneapi_kernel_compiler.asciidoc[
+sycl_ext_oneapi_kernel_compiler], which allows online compilation of a kernel
+from source code.
+These kernels must have a clean separation between host and device code, and
+they must have a defined order to their arguments because the application sets
+the argument values by their "index" via `handler::set_arg(index, value)`.
+Because variables captured by lambda expressions have no defined order, we need
```

Comment:
This is true for lambda expression but not for for non-lambda callable types. Isn't the motivation just that users might find this more convenient. If so why is the paragraph needed and shouldn't the only motivation be the next paragraph?

Reply:
I think there's some ambiguity even for a non-lambda callable.  For example, the spec doesn't make it clear whether an implementation passes the entire callable as a single argument or if the implementation passes each member variable as a separate kernel argument.  There's also no guarantee that the implementation passes all member variables, for example, in the case when some are not used.

Reply:
Why does the spec need to make this clear? It's up to the implementation to give correct results. And to that it needs to be consistent between device and host compiler. And in the case of different TU (/string) the host compiler can't see whether arguments are not used. So the only valid implementation is to pass them all. And it needs to use a consistent way to pass them as single/separate which doesn't require being aware of the content of the kernel (i.e. no heuristic which depends on the kernel).

Reply:
I think we decided in our phone conversation that this doesn't need further discussion, so I'm marking it resolved.